lib/inc/internal/substitution_expression.hpp

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <string>
+#include <memory>
+
+namespace hocon {
+
+    class substitution_expression : public std::enable_shared_from_this<substitution_expression> {
+    public:
+        substitution_expression(std::string path, bool optional);
+
+        std::string path() const;
+        bool optional() const;
+
+        std::shared_ptr<substitution_expression> change_path(std::string new_path);
+
+        std::string to_string() const;
+
+        template <typename T>
```

Comment:
This is to mimic the Java implementation which takes any type of `Object` to compare. I.e,
```
    @Override
    public boolean equals(Object other) {
        if (other instanceof SubstitutionExpression) {
...
```